### 2023.07.27
# 프로그램의 오류
- 어떤 원인에 의해 오독작 하거나 비정상적으로 종료되는 경우
- 발생시점에 따라 구분 가능
> - 컴파일 에러 (compile-time error) : 문법적 이슈 등의 이유로 런타임 시에 발생하는 에러
> - 런타입 에러 (runtime error) : 실행 과정에서 발생하는 에러, 문법적 오류 X
> - 논리적 에러 (logical error) : 의도와 다르게 나오는 것
# 에러와 예외
- Error : 발생하면 복구할 수 없는 심각한 오류
ex> 메모리 부족, 스택 오버 플로우
- Exception : 프로그래머가 적절한 코드를 통해 대비할 수 있는 오류
ex> 클래스 형변환 실패, 파일 읽기 실패
### 스택 오버 플로우란?
> 스택이라는 메모리 공간에서 재귀 호출을 너무 많이 하면 메모리가 초과되는 현상
# 예외처리(Exception Handing)
- 예외 발생 시 프로그램의 비 정상 종료를 막고 정상적인 실행 상태를 유지하는 것
- 예외의 감지 및 예외 발생 시 동작할 코드 작성 필요
# 예외의 계열
- RuntimeException 클래스들
> - 프로그래머의 실수로 발생하는 계열
> - unchecked expception
> - 대처 코드가 없어도 괜춘
- Exception 클래스들
> - 사용자의 실수 등 외적인 요인에 의해 발생하는 계열
> - checked exception
> - 대처 코드가 없으면 컴파일 진행 x
# 예외 처리 키워드
- 직접 처리
> try, catch, finally
- 간접 처리
> throws
- 사용자 정의 예외 발생시킬 때
> throw
# try~catch(직접처리)
- 프로그램 실행 시 발생할 수 있는 예외에 대한 대비코드 작성
- 비정상적인 종료를 막고, 정상적인 상태를 유지함.
- 예외 발생시 JVM의 예외처리기가 닫아서 처리  
예외 발생 시 실행코드 : 1->3->4  
예외가 발생하지 않았을 때 실행코드 : 1->2->4  
예외를 처리하지 못 했을 때 실행코드 : 1(중단)
```java
try {
  // (1) 예외가 발생할 수 있는 코드
  // (2) 정상코드
} catch (Exception e) {
  // (3) 예외가 발생했을 때 처리할 코드
}
// (4) 일반코드
```

- **중괄호(블록) 생략 불가능**
- 블럭안에 또다른 try~catch 구문이 올 수 있음.
# 다중 예외처리
- try 블록에서 여러 종류의 예외가 발생할 경우
- 하나의 try 블록에 여러 개의 catch 블록 추가 가능
```java
// 위에서부터 차례대로 확인
try {
  // exception 발생할만한 코드
} catch (XXException e) {
  // XXException 발생 시 처리 코드
} catch (YYException e) {
  // YYExpception 발생 시 처리 코드
}
```
```java
try {
  // exception이 발생할만한 코드
} catch (XXException | YYException e) {
  // 필요시 instanceof를 통해 나누어 작성
  // XXException 발생 시 처리 코드
  // YYException 발생 시 처리 코드
}
```
# 다중 예외처리 유의 사항
- JVM이 던진 예외는 catch 문장을 찾을 때는 다형성이 적용됨
- 상위 타입의 예외가 먼저 선언되는 경우 뒤에 등장하는 catch 블록은 동작할 기회가 없음
- 상속 관계가 없는 경우는 무관
- **상속 관계에서는 작은 범위(자식)에서 큰 범위(조상)순으로 정의**
```java
try {
  // exception 발생할만한 코드
} catch (XXException e) {
  // XXException 발생 시 처리 코드
} catch (YYException e) {
  // YYExpception 발생 시 처리 코드
// 만능처리기 Exception아 부탁해~!
// Exception은 맨마지막에 와야한다
} catch (Exception e) {
  // Exception 발생 시 처리 코드
}
```
# Exception 인스턴스의 주요 메서드
- getMessage() : 발생된 예외에 대한 구체적인 메시지를 반환
- printStackTrace() : 예외 발생 당시의 호출 스택(Call Stack)을 출력한다.
# 메서드에서 예외 던지기(간접처리)
- 메서드 선언부에 **throws** 키워드를 사용하여 예외 작성
- 예외가 여러 개일 경우 **,**를 아용하여 나영
- 메서드가 예외를 처리하는 것이 아닌 **전달**
- 메서드 : 난 처리 안해! 던질게 부탁해~
- 조상 타입의 예외로 처리 가능
# checked exception과 throws
![ex_screenshot](/images/checked_exception과_throws.PNG)
# runtime exception과 throws
![ex_screenshot](/images/runtime_exception과_throws.PNG)
# 메서드 재정의와 throws
- 메서드 재정의 시 조상클래스 메서드가 던지는 예외보다 부모 예외를 던질 수 없다.
![ex_screenshot](/images/메서드_재정의와_throws.png)  
이거 안됨
# finally 구문
- finally는 예외 발생 여부와 상관 없이 언제나 실행
```java
try {
  // 예외가 발생할 가능성이 있는 코드
}catch(Exception e){
  // 예외 처리 코드
}finally {
  // 예외 상관없이 항상 수행해야 하는 코드
}
```
- 예외 발생시 -> try-catch-finally 순으로 코드 실행
- 예외 미발생시 -> try-finally 순으로 코드 실행
- **중간에 return이 있어도 finally 블록 수행 후 반환**
```java
	// finally는 무조건 실행!!!
	// 1. 정상적으로 실행시 1 2 4 5
	// 2. 예외 발생시 1 3 4 5
	// 3. 예외 미처리시 1 4(중단)
	// 4. try 구문에 return이 들어가면 (정상 시행시) 1 2 4
	// 5. catch 구문에 return이 들어가면 (예외 발생시) 1 3 4
	public static void main(String[] args) {
		int num = 0;
		int[] nums = {1};
		try {
			System.out.println("1");
			int i = 1/num;
//			nums[2] = 10;
			System.out.println("2");
//			return;
		} catch (ArithmeticException e) {
			System.out.println("3");
			return;
		} finally {
			System.out.println("4");
		}
		System.out.println("5");
	}
```
# 자동 자원 반납 구문(try with resources)
- 자원 등을 반납할 때 finally에서 close()를 통해 반납을 주로 함.
- 코드가 지저분해지고 다른 예외상황을 발생시킬 수 있음.
```java
FileInputStream fis = null;
try {
  fis = new FileInputStream("test.txt");
} catch (IOException e) {
  e.printStackTrace();
} finally {
  try {
    if (fis != null)
      fis.close();
  } catch (IOException e) {
    e.printStackTrace();
  }
}
```
- try() -> 괄호 안에 객체를 생성하는 코드를 작성하면, 해당 객체는 close()를 호출하지 않아도
  블록을 벗어 나는 순간 close()가 호출됨.
- 해당 객체의 클래스가 AutoCloeseable이라는 인터페이스를 구현한 클래스이어야만 함.
```
try (FileInputStream fis = new FileInputStream("test.txt")) {
  // 코드생략
} catch (IOException e) {
  e.printStackTrace();
}
```
# 사용자 정의 예외
- 기존에 정의된 예외 이외에 사용자가 직접 정의 예외를 작성할 수 있음
- 대부분 Exception 또는 RuntimeException 클래스를 상속받아 작성
> - checked exception 활용 : 빨간줄로 눈에 보여! 명시적 예외 처리 또는 throws 필요  
    (코드는 복잡해지지만 처리, 누락 등 오류 발생 가능성은 down)
> - runtime exception 활용 : 묵시적 예외 처리 가능  
    (코드가 간결해지지만 예외 처리, 누락 가능성 발생)
